### PR TITLE
Adding apidiff target to the Makefile

### LINKED ISSUE #3169: Add apidiff target to the Makefile
We should add a Make target that takes a diff on API changes and create a optional pre-submit job that calls that target. 

We can follow what CAPI already has:
Adding a [Makefile target to build apidiff](https://github.com/kubernetes-sigs/cluster-api/blob/main/Makefile#L746-L747)
Adding a [Makefile target to run apidiff locally against origin/main](https://github.com/kubernetes-sigs/cluster-api/blob/535219878ced6ce483c87fe49b4a756c0f83a85a/Makefile#L310)
Adding an optional ProwJob to run apidiff in CI: [here](https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes-sigs/cluster-api/cluster-api-presubmits-main.yaml#L21-L39)

This will improve our confidence on not introducing breaking API changes.

### PR DESCRIPTION
**Pull request by meghanajangi**
Signed-off-by: Meghana Jangi <mjangi@vmware.com>

**What type of PR is this?**

**What this PR does / why we need it**:
Adds a Make target that takes a diff on API changes.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3169 

**Special notes for your reviewer**:

**Checklist**:

- [x] squashed commits
- [ ] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests

**Release note**:

```release-note
Add apidiff Make target
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -629,3 +634,10 @@ clean-temporary: ## Remove all temporary files and folders
 	rm -rf test/e2e/capi-kubeadm-control-plane-controller-manager
 	rm -rf test/e2e/logs
 	rm -rf test/e2e/resources
+
+##@ hack/tools:
+
+go-apidiff: $(GO_APIDIFF) ## Build a local copy of apidiff
+
+$(GO_APIDIFF): $(TOOLS_DIR)/go.mod # Build go-apidiff from tools folder.
+	cd $(TOOLS_DIR) && go build -tags=tools -o $(GO_APIDIFF_BIN) github.com/joelanford/go-apidiff
```
Comment (Ankitasw) [inline Makefile:643]: nit: newline required here

File: hack/tools/go.mod [hunk 2]
```diff
@@ -8,6 +8,7 @@ require (
 	github.com/golang/mock v1.6.0
 	github.com/golangci/golangci-lint v1.44.0
 	github.com/itchyny/gojq v0.12.6
+	github.com/joelanford/go-apidiff v0.1.0
```
Comment (Ankitasw) [inline hack/tools/go.mod:11]: There is v0.2.0 available, do we want to use that?

File: Makefile (hunk 1 above)
Comment (meghanajangi) [reply Makefile]: Added newline as suggested

**Comment by Ankitasw** [comment]
@meghanajangi have you locally tested `make apidiff`. It gives error for me.

> cd hack/tools && go build -tags=tools -o bin/go-apidiff github.com/joelanford/go-apidiff
go: downloading golang.org/x/exp v0.0.0-20211029160041-3396431c207b
hack/tools/bin/go-apidiff 33c9ef97f852bf9b0866977911184e4e0f651ba5 --print-compatible
git tree is dirty
make: *** [apidiff] Error 2

**Comment by meghanajangi** [comment]
> @meghanajangi have you locally tested `make apidiff`. It gives error for me.
> 
> > cd hack/tools && go build -tags=tools -o bin/go-apidiff github.com/joelanford/go-apidiff
> > go: downloading golang.org/x/exp v0.0.0-20211029160041-3396431c207b
> > hack/tools/bin/go-apidiff [33c9ef9](https://github.com/kubernetes-sigs/cluster-api-provider-aws/commit/33c9ef97f852bf9b0866977911184e4e0f651ba5) --print-compatible
> > git tree is dirty
> > make: *** [apidiff] Error 2

I have fixed the git tree issue now, can you please check again?

**Comment by Ankitasw** [comment]
@meghanajangi still getting it in my local, do you run any other command before, I cleaned up the generated files as well.

File: Makefile [hunk 3]
```diff
@@ -629,3 +634,10 @@ clean-temporary: ## Remove all temporary files and folders
 	rm -rf test/e2e/capi-kubeadm-control-plane-controller-manager
 	rm -rf test/e2e/logs
 	rm -rf test/e2e/resources
+
+##@ hack/tools:
+
+go-apidiff: $(GO_APIDIFF) ## Build a local copy of apidiff
+
+$(GO_APIDIFF): $(TOOLS_DIR)/go.mod # Build go-apidiff from tools folder.
```
Comment (sedefsavas) [inline Makefile:642]: How about moving this to hack/tools?

Comment (sedefsavas) [reply Makefile:642]: Just checked, we already have this target there:

https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/24e3ea2e372e5c6deff3c32bb6e72e4207f0f24a/hack/tools/Makefile#L67-L68

Comment (meghanajangi) [reply Makefile]: Removed these lines from the change.

**Pull request merged**